### PR TITLE
Import disambiguation: Corrects handling of fully-applied and one-sided sectioned operators in qualifying strategy

### PR DESCRIPTION
**Pull request by konn**
The current implementation of the import disambiguation (introduced in #1264) always qualifies operators with extra parens.
This can destroy the fully- or partially-applied infix operators, as follows:

```hs
op = (++)
fun xs ys = xs ++ ys
funL xs = (++ xs)
funR ys = (ys ++)
```

↓

```hs
op = (Prelude.++) -- As expected
fun xs ys = xs (Prelude.++) ys -- oops!
funL xs = ((Prelude.++) xs)  -- ooops!
funR ys = (ys (Prelude.++)) -- oooops!
```

This PR just fixes it, inspecting the buffer whether the code in question is already parenthesised or not.

### CODE REVIEW

**Review by jneira — APPROVED** [review]
lgtm, and good test coverage

File: ghcide/src/Development/IDE/Plugin/CodeAction.hs [hunk 1]
```diff
@@ -832,14 +839,19 @@ disambiguateSymbol pm Diagnostic {..} (T.unpack -> symbol) = \case
                     else hideSymbol symbol <$> imps
                 | ImplicitPrelude imps <- hiddens0
                 ]
-    (ToQualified qualMod) ->
+    (ToQualified parensed qualMod) ->
         let occSym = mkVarOcc symbol
             rdr = Qual qualMod occSym
-         in [ Rewrite (rangeToSrcSpan "<dummy>" _range) $ \df -> do
-                liftParseAST @(HsExpr GhcPs) df $
+            asExpr = parensed && isSymOcc occSym
```
Comment (berberman) [inline ghcide/src/Development/IDE/Plugin/CodeAction.hs:845]: Can we remove `isSymOcc` here?

**Comment by berberman** [comment]
If you don't mind, could you unify the filename extension of golden tests data to `.hs`? It would be better for review and can clarify that our codebase doesn't contain E language source code :smile:

**Comment by Ailrun** [comment]
Yes, especially when the "expected" version is better source, like in this case.

**Comment by konn** [comment]
Ah, I just followed the pattern used in Eval plugin, but, yes, those `expected` files in eval func-test doesn't have so much long extensions. I now renamed disambiguation-related test files to have `.hs` extensions. Thank you for your suggestions, @berberman @Ailrun!

**Comment by berberman** [comment]
> Can we remove isSymOcc here?

Did you see my comment? I think it's enough to consider only `parensed`. If a non-symbolic identifier is parenthesized unnecessarily, the range reported in diagnostic won't include the parenthesis part, i.e. `parensed` will be `False`, so I think it doesn't matter if `isSymOcc occSym` is `True`. Correct me if I'm wrong :)

**Comment by konn** [comment]
> > Can we remove isSymOcc here?
> 
> Did you see my comment? I think it's enough to consider only `parensed`. If a non-symbolic identifier is parenthesized unnecessarily, the range reported in diagnostic won't include the parenthesis part, i.e. `parensed` will be `False`, so I think it doesn't matter if `isSymOcc occSym` is `True`. Correct me if I'm wrong :)

Oops, sorry I overlooked your comment 🙇 I think your suggestion makes sense. Thanks!

File: ghcide/src/Development/IDE/Plugin/CodeAction.hs (hunk 1 above)
Comment (konn) [reply ghcide/src/Development/IDE/Plugin/CodeAction.hs]: I've just changed it to use `parensed` only. Thanks!

**Comment by jneira** [comment]
very happy to see more contributors with the commit bit, congrats and thanks for your great work here :slightly_smiling_face:

**Pull request merged**